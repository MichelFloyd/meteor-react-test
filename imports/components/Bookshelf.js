import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { injectSheet } from '/imports/styling';
import Book from './Book';

const styles = {
  root: {
    display: 'flex',
    flexWrap: 'wrap',
  },
};

export const BookshelfHelper = Wrapped => {
  return class extends Component {
    constructor(props) {
      super(props);
      this.state = {
        selectedBookIds: [],
      };
    }

    bookSelected = bookId => {
      return this.state.selectedBookIds.indexOf(bookId) >= 0;
    };

    toggleBookSelection = bookId => {
      let newSelection = this.state.selectedBookIds;
      newSelection = this.bookSelected(bookId) ? newSelection.filter(id => id !== bookId) : [...newSelection, bookId];
      this.setState({ selectedBookIds: newSelection });
    };

    clearBookSelection = () => {
      this.setState({ selectedBookIds: []});
    }

    render() {
      return (
        <Wrapped
          {...this.props}
          selectedBookIds={this.state.selectedBookIds}
          onToggleBookSelection={this.toggleBookSelection}
          onClearBookSelection={this.clearBookSelection}
        />
      );
    }
  };
};

const Bookshelf = ({
  books,
  classes,
  onToggleBookSelection,
  selectedBookIds,
}) => {
  return (
    <div className={'bookshelf ' + classes.root}>
      {books.map( book => (
        <Book
          key={book.etag}
          book={book}
          selected={selectedBookIds.indexOf(book.etag) >= 0}
          onClick={onToggleBookSelection.bind(this,book.etag)}
        />
      ))}
    </div>
  );
};

Bookshelf.propTypes = {
  books: PropTypes.arrayOf(PropTypes.object).isRequired,
  onToggleBookSelection: PropTypes.func.isRequired,
  selectedBookIds: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default injectSheet(styles)(Bookshelf);
