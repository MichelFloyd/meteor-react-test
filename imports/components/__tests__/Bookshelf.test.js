import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import Bookshelf from '../Bookshelf';
import books from './fake-book-data';

const selectionFn = jest.fn();

test('clicking a book adds it to the selection', () => {
  const bookshelfJSX = (
    <Bookshelf
      books={books}
      onToggleBookSelection={selectionFn}
      selectedBookIds={[]}
    />
  );
  const component = renderer.create(bookshelfJSX);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const wrapper = mount(bookshelfJSX);
  const bookshelf = wrapper.find('.bookshelf');
  bookshelf.childAt(0).simulate('click');

  expect(selectionFn.mock.calls.length).toBe(1);
  expect(selectionFn.mock.calls[0][0]).toBe('abc123');
});
