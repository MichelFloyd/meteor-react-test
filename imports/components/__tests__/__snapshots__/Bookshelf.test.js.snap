// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clicking a book adds it to the selection 1`] = `
<div
  className="bookshelf root-0-0"
>
  <div
    className="root-0-1"
    onClick={[Function]}
  >
    <div>
      <div
        className="root-0-3"
        style={
          Object {
            "height": 187,
            "width": 128,
          }
        }
      >
        <img
          className="image-0-6"
          height={187}
          src="http://placehold.it/128x187"
          title="Not a real book"
          width={128}
        />
        <div
          className="reflection-0-7"
        >
          <img
            className="flippedImage-0-4"
            height={187}
            src="http://placehold.it/128x187"
            width={128}
          />
          <div
            className="reflectionOverlay-0-9"
            style={
              Object {
                "height": 187,
                "width": 128,
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="root-0-1"
    onClick={[Function]}
  >
    <div>
      <div
        className="root-0-3"
        style={
          Object {
            "height": 187,
            "width": 128,
          }
        }
      >
        <img
          className="image-0-6"
          height={187}
          src="http://placehold.it/128x187"
          title="Also not a real book"
          width={128}
        />
        <div
          className="reflection-0-7"
        >
          <img
            className="flippedImage-0-4"
            height={187}
            src="http://placehold.it/128x187"
            width={128}
          />
          <div
            className="reflectionOverlay-0-9"
            style={
              Object {
                "height": 187,
                "width": 128,
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
